### ISSUE TITLE
lifespan events in sub-applications

### ISSUE THREAD

**Issue by octopoulpe** [Kludex/starlette#649]
Hi.
I was wondering if sub-applications (mounted on the main application) should receive startup/shutdown events and call their own handlers.
I saw it is not the case at the moment. Should we document it or implement the dispatching of lifespan events ?

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/649">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/649/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/starlette/issues/649/pledge.svg">
</picture>
</a>

**Comment by gbozee** [comment]
Assuming the sub applications are connected using the Router, startup/shutdown events can be implemented as follows

```python
app = Router([Mount("/api", app=service_app), Mount("", app=root_app)])


async def open_database_connection_pool():
    await service_app.service.connect()


async def close_database_connection_pool():
    await service_app.service.disconnect()


app.lifespan.add_event_handler("startup", open_database_connection_pool)
app.lifespan.add_event_handler("shutdown", close_database_connection_pool)

```

**Comment by octopoulpe** [comment]
Hi.
I chose to manage that through the main application handlers.
```
app = Starlette()

sub_applications = {"/cart": cart_app, "/account": account_app, etc...}
for route, sub_app in sub_applications.items():
    app.mount(route, sub_app)
    
@app.on_event("startup")
async def startup():
    for sub_app in sub_applications.values():
        await sub_app.router.lifespan.startup()
```

The same for shutdown. As long as you do not go all 'subsubapplication', I think it is quiet expressive as, in my case, sub application don't know they are "mounted" somewhere.

Shouldn't we, at least, document it ?

**Comment by lovelydinosaur** [comment]
> I was wondering if sub-applications (mounted on the main application) should receive startup/shutdown events and call their own handlers.

Ideally, yes.

**Comment by lovelydinosaur** [comment]
Having taken a look at this it's really quite awkward.

You can't run each submount off the same lifespan task - instead each would need to have their own sub-task. Then you need to make sure you're handling the case where lifespan events are not supported and deal with exception cases from each submount.

At the moment I think we need to just treat it as out-of-scope.

But yeah, we do need to document that.

**Comment by sm-Fifteen** [comment]
Having applications that may or may not behave properly depending on whether they're mounted at the root of the server or as a sub-application due to sub-apps not receiving lifespan events is probably going to cause issues if using #799 context-managers becomes the recommended way of handling application-level resources instead of global scope [like the doc currently advises](https://www.starlette.io/database/). It's one of my main worries in regards to tiangolo/fastapi#617, because it makes applications that rely on lifetime events for initialization non-composable.

Does that issue look any less awkward to fix now than it used to?

**Comment by adriangb** [comment]
I think that if we deprecated the non-contextmanager lifespans we should be able to implement lifespan propagation.

I see 2 ways to go about it:

1. The top-level router (which receives the ASGI lifespan event) "crawls" all of it's routes, submounts, etc. This can technically be done with no API changes since `Router.lifespan` is a public attribute and so are `Router.routes`, `Mount.app`, etc.
2. We add an `async def lifespan(self, app: ASGIApp) -> AsyncContextManager[None]` method to every routing class and have them propagate the lifespan.

TLDR when propagating the lifespans, skip the whole ASGI aspect of of it and instead just call the context managers directly.

**Comment by adriangb** [comment]
Coming back to this a couple months later. I don't agree with myself above anymore. I think it _is_ possible to do this in pure ASGI. That is not to say that introspecting into routers and mounts (and maybe adding a public `lifespan()` context manger) _won't_ work, it will. But I've also been able to get this to work with pure ASGI.

Some concrete examples to look at:
1. Introspection-based. [Xpresso currently uses this approach](https://github.com/adriangb/xpresso/blob/main/xpresso/applications.py#L239-L289). The main con with this approach is that whatever is introspecting needs to have knowledge of the routing classes. If everything is coming from Starlette, it kinda just works. But if you mount a non-Starlette app, this will not work.
2. Pure ASGI. It's tricky, but I think it's possible to write an ASGI app that "propagates" lifespan events. I packaged this approach up as a middleware, but I think the Starlette application could do the same thing: https://github.com/adriangb/asgi-lifespan/blob/main/asgi_lifespan_middleware/_middleware.py. It's certainly not _simple_ code, but I think it works (please poke holes in it if you can). The huge pro of this is that it works with _any_ ASGI app. But what about routers? A router is a 1 to many mapping from a containing app to contained routes. Those routes can be mounted apps. So we don't just need to run the lifespan of a single mounted app: we need to run lifespans for _multiple_ mounted apps. Surprisingly, a similar approach seems to work: https://github.com/adriangb/asgi-routing/pull/1/files#diff-16a5f890ec0ac83174fb209a886f9149b9ac775a62254d5f958737701565601e.

At this point, it seems like a tradeoff between complexity and functionality. The simplest thing to do is just not support lifespans for mounted apps at all. The least complex (in terms of LOC at least) would be to introspect into the routing system and pull-out lifespans from routing primitives we recognize. The most complex version would be where `Mount` does something like [this](https://github.com/adriangb/asgi-lifespan/blob/main/asgi_lifespan_middleware/_middleware.py) and `Router` does something like [this](https://github.com/adriangb/asgi-routing/pull/1/files#diff-16a5f890ec0ac83174fb209a886f9149b9ac775a62254d5f958737701565601e).

**Comment by chickeaterbanana** [comment]
I tried a solution as well. The solution seems to work quite well as far as I can see. I solved following challenges for me:
* support event and context at the same time
* support lifespan propagation
* support multiple lifespan contextmanagers

The solution can be found here: https://github.com/encode/starlette/commit/bdc9026b642c16e540f2d2d58e43edd927e7062b

**Comment by verigle** [comment]
> ```python
> app = Router([Mount("/api", app=service_app), Mount("", app=root_app)])
> 
> 
> async def open_database_connection_pool():
>     await service_app.service.connect()
> 
> 
> async def close_database_connection_pool():
>     await service_app.service.disconnect()
> 
> 
> app.lifespan.add_event_handler("startup", open_database_connection_pool)
> app.lifespan.add_event_handler("shutdown", close_database_connection_pool)
> ```

how to import Router?

**Comment by mattmess1221** [comment]
Here's a lifespan I wrote as a workaround to this. It supports both `Host` and `Mount` route types.

```py
@contextlib.asynccontextmanager
async def mounted_lifespan(app: Starlette) -> AsyncGenerator[Mapping[str, Any], Any]:
    async with contextlib.AsyncExitStack() as stack:
        state = {}
        for r in app.routes:
            if isinstance(r, Host | Mount) and isinstance(r.app, Starlette | Router):
                router = r.app
                if isinstance(router, Starlette):
                    router = router.router
                result = await stack.enter_async_context(router.lifespan_context(r.app))
                if result is not None:
                    state.update(result)

        yield state
```

Caveats:
- It only supports `Starlette` or `Router` mount types.
- All state is merged into the root app, which may cause conflicts.
- Will not trigger middleware lifespan. It would cause the event `lifespan.startup.complete` to be sent multiple times.

Full example:
```py
# mounted app's lifespan
@contextlib.asynccontextmanager
async def test_lifespan(app: Starlette) -> AsyncGenerator[dict, Any]:
    print("Hello, lifespan")
    yield {"data": "stuff"}


# mounted app endpoint which depends on state created by the lifespan
async def test_endpoint(request: Request) -> Response:
    return PlainTextResponse(request.state.data)


# child app with lifespan
test_app = Starlette(
    routes=[
        Route("/", test_endpoint),
    ],
    lifespan=test_lifespan
)

# root app with mounted apps
app = Starlette(
    routes=[
        Mount("/test_app", test_app),
    ],
    lifespan=mounted_lifespan,
)
```

**Comment by mattmess1221** [comment]
Converting it to a middleware solves caveat 1. Doing some magic with `asyncio.Queue` and wrapping send/receive solves caveat 3.

```py
class MountedLifespanMiddleware:
    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> Any:
        if scope["type"] == "lifespan":
            queues: list[Queue[Message]] = []

            async def wrap_receive() -> Message:
                msg = await receive()
                async with anyio.create_task_group() as tasks:
                    for queue in queues:
                        tasks.start_soon(queue.put, msg)
                return msg

            async def wrap_send(message: Message) -> None:
                if message["type"] == "lifespan.startup.complete":
                    return
                await send(message)

            async with anyio.create_task_group() as tasks:
                app = scope.get("app")
                if isinstance(app, Starlette):
                    for r in app.routes:
                        if isinstance(r, Mount | Host):
                            queues.append(queue := Queue())
                            tasks.start_soon(r.app, scope, queue.get, wrap_send)

                await self.app(scope, wrap_receive, send)
                return

        await self.app(scope, receive, send)
```

**Comment by Cyberbolt** [comment]
Solution:

```python
from fastapi import FastAPI
from contextlib import asynccontextmanager, AsyncExitStack

sub_app = FastAPI()

@asynccontextmanager
async def sub_lifespan(app: FastAPI):
    print("sub app is starting up...")
    yield
    print("sub app is shutting down...")

# Use AsyncExitStack to manage context
async def lifespan(app: FastAPI):
    async with AsyncExitStack() as stack:
        # Manage the lifecycle of sub_app
        await stack.enter_async_context(sub_lifespan(sub_app))
        yield

app = FastAPI(lifespan=lifespan)

# Mount the sub application
app.mount("/sub", sub_app)

# Define the main application's route
@app.get("/")
async def read_main():
    return {"message": "This is the main app"}

# Define the sub application's route
@sub_app.get("/")
async def read_unipay():
    return {"message": "This is the unipay app"}

**Comment by racinmat** [comment]
Any progress on this?
There are new use-cases that I think are relevant: in AI Agents, both a2a and adk frameworks have FastAPI integration and use lifespans:

https://github.com/a2aproject/a2a-python/blob/v0.2.14/src/a2a/server/apps/jsonrpc/fastapi_app.py#L131
https://github.com/google/adk-python/blob/v1.7.0/src/google/adk/cli/fast_api.py#L281
Being able to mount them to existing application would be great help.

**Comment by racinmat** [comment]
If anyone is interested, I made a more generic workaround:
```python

def extract_lifespan_from_app(app) -> Callable[[Any], AsyncContextManager[None]] | None:
    """
    Extract lifespan function from a FastAPI/Starlette application.
    Args:
        app: FastAPI or Starlette application instance
    Returns:
        The lifespan function if it exists, None otherwise
    """
    if hasattr(app, 'router') and hasattr(app.router, 'lifespan_context'):
        # FastAPI/Starlette with lifespan
        lifespan_context = app.router.lifespan_context
        if lifespan_context is not None:
            return lifespan_context

    # Check if app has direct lifespan attribute (older FastAPI versions)
    if hasattr(app, 'lifespan') and app.lifespan is not None:
        return app.lifespan

    return None


def find_all_mounted_apps(app: FastAPI, path_prefix: str = "") -> list[tuple[str, Any]]:
    """
    Recursively find all mounted applications in a FastAPI app.
    Args:
        app: The main FastAPI application
        path_prefix: Current path prefix for nested mounts
    Returns:
        List of tuples (mount_path, app_instance)
    """
    mounted_apps = []   # not using the app itself, because we're calling it inside the lifespan itself

    if not hasattr(app, 'router') or not hasattr(app.router, 'routes'):
        return mounted_apps

    for route in app.router.routes:
        if isinstance(route, Mount):
            mount_path = path_prefix + route.path
            mounted_app = route.app

            # Add this mounted app
            mounted_apps.append((mount_path, mounted_app))

            # Recursively check for nested mounts
            if hasattr(mounted_app, 'router'):
                nested_apps = find_all_mounted_apps(mounted_app, mount_path)
                mounted_apps.extend(nested_apps[1:])  # Skip the root app we already added

    return mounted_apps


def extract_all_lifespans(app: FastAPI) -> list[tuple[Callable[[Any], AsyncContextManager[None]], FastAPI]]:
    """
    Extract all lifespan functions from the main app and all mounted sub-apps.
    Args:
        app: The main FastAPI application
    Returns:
        List of lifespan functions found in the app hierarchy
    """
    lifespans = []
    mounted_apps = find_all_mounted_apps(app)

    for mount_path, mounted_app in mounted_apps:
        lifespan = extract_lifespan_from_app(mounted_app)
        if lifespan is not None:
            lifespans.append((lifespan, mounted_app))

    return lifespans


@asynccontextmanager
async def composed_lifespan(app: FastAPI) -> AsyncGenerator[None, Any]:
    """
    Natively, FastAPI lifespan does not get propagated to sub applications
    https://fastapi.tiangolo.com/advanced/events/#sub-applications
    but we desire that behavior, because some apps rely on it.
    """
    lifespans = [(<lifespan that you would like to use in the app originally>, app)] + extract_all_lifespans(app)
    async with AsyncExitStack() as stack:
        for a_lifespan, an_app in lifespans:
            await stack.enter_async_context(a_lifespan(an_app))
        yield

app = FastAPI(lifespan=composed_lifespan)
```
and this works and automatically discovers lifespan of mounted sub-app.

**Comment by racinmat** [comment]
Hm, this https://github.com/fastapi/fastapi/issues/811#issuecomment-1870030103 looks even better.

**Comment by IgnazioDS** [comment]
The MCP use case mentioned at the bottom of this thread is now one of the most common real-world triggers for this issue. When you mount an MCP server (`from mcp.server.fastmcp import FastMCP`) onto an existing FastAPI app, the MCP server uses its own lifespan to initialize database connections, cached metadata, or tool registries. Without lifespan propagation, those resources never initialize, and you get 500 errors as soon as the first tool call hits.

The `racinmat` workaround with `composed_lifespan` is the cleanest approach I’ve found, but it requires explicit knowledge of all sub-apps at composition time. For plugin-style architectures (where sub-apps are registered at startup, not at import time), `racinmat`'s `extract_all_lifespans` recursive discovery approach works, with one addition — you need to call `extract_all_lifespans` *after* all `app.mount(...)` calls, not at module import time.

For the specific MCP case, here’s the minimum working pattern:

```python
from contextlib import asynccontextmanager, AsyncExitStack
from fastapi import FastAPI
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("my-tools")
mcp_app = mcp.get_application()  # returns an ASGI app with its own lifespan

@asynccontextmanager
async def lifespan(app: FastAPI):
    async with AsyncExitStack() as stack:
        # Manually include the MCP app's lifespan
        if hasattr(mcp_app, 'router'):
            await stack.enter_async_context(mcp_app.router.lifespan_context(mcp_app))
        yield

app = FastAPI(lifespan=lifespan)
app.mount("/mcp", mcp_app)
```

The deeper fix in Starlette would be to have `Mount` automatically participate in lifespan propagation when the mounted app is itself a Starlette/FastAPI instance — `adriangb`'s `MountedLifespanMiddleware` approach or something equivalent baked into `Router.lifespan_context` seems like the right long-term direction. Happy to see this land in Version 1.x.
